c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1203,14 +1286,251 @@ class DeviceCachingAllocator {
     reset_peak_stat(stats.oversize_segments);
   }
 
+  /* Checkpoint the state of a private pool necessary to return it to its
+   * current state */
+  PrivatePoolState getCheckpointState(MempoolId_t id) {
+    std::lock_guard<std::recursive_mutex> lock(mutex);
+
+    auto pool = graph_pools.find(id);
+    if (pool != graph_pools.end()) {
+      auto private_pool_blocks = get_private_pool_blocks(pool->second.get());
+      return constructPrivatePoolState(
+          pool->second.get(), id, private_pool_blocks);
+    } else if (graph_pools_freeable.count(id)) {
+      TORCH_CHECK(false, "Not expected to checkpoint freeable graph");
+    } else {
+      TORCH_CHECK(false, "Could not find pool of id");
+    }
+  }
+
+  void freeBlocksAllocatedToPool(PrivatePool* private_pool) {
+    std::unordered_map<void*, Block*> orig_ptrs_to_blocks;
+
+    auto pool_blocks = get_private_pool_blocks(private_pool);
+
+    std::vector<Block*> head_blocks;
+    for (Block* block : pool_blocks) {
+      if (block->prev == nullptr) {
+        head_blocks.push_back(block);
+      }
+    }
+
+    for (Block* block : head_blocks) {
+      Block* curr = block;
+
+      while (curr) {
+        // When we free a block, its pointer should never change
+        // only its adjacent blocks, so free, then look at pointer
+        if (curr->allocated) {
+          TORCH_CHECK(
+              curr->event_count == 0,
+              "Events should have synchronized when setting checkpointed block");
+          local_raw_delete(curr->ptr);
+          TORCH_CHECK(!curr->allocated)
+        }
+        curr = curr->next;
+      }
+    }
+
+    for (Block* b : get_private_pool_blocks(private_pool)) {
+      TORCH_CHECK(!b->allocated);
+    }
+  }
+
+  // checkpoint the state of an allocation that may have been
+  // split into multiple blocks
+  void checkpointSegmentState(
```

Comment:
I can't quite follow this function. I think it is basically restoring a segment to a list of allocated/unallocated blocks but the pre and post conditions are not clear. I think that there is a potential bug: allocate_found_block uses `should_split` logic whose behavior can change based on allocator settings that maybe have changed since the checkpoint. It might be safer for this function to just directly reconstitute the blocks then couple it to the behavior of how blocks are are managed on the alloc path. A function like:

```
RestoreResult {
   std::vector<void*> allocations_freed;
   std::vector<Block*> allocations_created;
};
RestoreResult restoreSegment(SegmentState state);
```

which can then be used in native allocator to fix up the state after the restore.

Reply:
Great point about the bug. One possibility here is just to add a check after checkpointing that the segment is in the state it was checkpointed to be. The advantage of that is that I don't have to as closely recreate all of the stat information and some of the other logic that is contained elsewhere. WDYT ?